### ISSUE TITLE
Want minimal sftp function

### ISSUE THREAD

**Issue by David263** [phpseclib/phpseclib#1564]
While I appreciate the great range of algorithms implemented in phpseclib, I would like to find a short function that does an sftp file upload only, given account name, password, and the private key and certificate file pathnames. So far, web searching has not produced anything. Any ideas?

**Comment by malohtie** [comment]
Hi @David263 

You can find everything on the official website

https://phpseclib.com/docs/auth

https://phpseclib.com/docs/sftp

lmk if you need more help

good luck

**Comment by David263** [comment]
Thank you, and happy holidays!

Is phpseclib already part of PHP 7? Or does it need to be downloaded? Is there an introduction somewhere that answers such questions?

**Comment by malohtie** [comment]
> Thank you, and happy holidays!

You to


phpseclib support php 7, it a package that can be downloaded and installed via composer

to install it choose a version that you need and follow those steps 

https://github.com/phpseclib/phpseclib#branches

if you don't have composer you can get it from here https://getcomposer.org/

**Comment by terrafrost** [comment]
> I would like to find a short function that does an sftp file upload only, given account name, password, and the private key and certificate file pathnames

Unless you're using multi factor authentication or encrypted private keys you probably don't need _both_ the password and private key.

And what are you talking about w.r.t. the certificate? [RFC6187](https://tools.ietf.org/html/rfc6187) discusses using X.509 certs for signatures but it doesn't seem widely used at all. At the very least phpseclib doesn't support it. Is that what you're referring to or are you confusing SSL/TLS with SSH or something?

In lieu of knowing those answers I'd suggest building off of what malohtie has posted and build your own method. eg.

```
ssh_upload($host, $user, $pass, $remote_file_name, $local_file_name)
```
I mean, if `$remote_file_name` lives in the current directory and `$local_file_name` is supposed to be the same and you want the script to upload it to the default directory (presumably the /home directory of the logged in user) then you can omit the second filename and just leave the first idk lol

**Comment by David263** [comment]
Thank you, everyone. I know very little about how sFTP works, except that it is  built on SSH. I am fine using encrypted or unencrypted private key and don't need to use the password. This is what I want to use to copy development files up to my remote server. I am certain of that. Not FTPS or the hundreds of other methods. I want to use what PuTTY uses and I want to find a way to turn off plain FTP access to my remote server. Why? For good security during uploading.

I don't want to download anything. I don't need large libraries like phpseclib that are not already a part of Apache or PHP (I'm assuming it is large since I cannot find it to see its size).

I just want some minimal PHP code to do the uploading. Surely someone has posted this somewhere on the Web but I cannot find it. I found some functions in the PHP Manual, but no clear example of usage like there is for insecure FTP (using ftp_connect, ftp_login, ftp_put, and ftp_close).

**Comment by David263** [comment]
Here is part of what I have created on my own so far:

`$methods=['hostkey'=>"ssh-rsa"];
$ftp=ssh2_connect($c['host'],$c['securePort'],$methods);
`
The error message is:

`Call to undefined function ssh2_connect() `

Do I need to edit PHP.ini?

**Comment by terrafrost** [comment]
> (I'm assuming it is large since I cannot find it to see its size)

If you're using cPanel then you're on a shared host so the best version for you would be 1.0, which can be downloaded here:

https://sourceforge.net/projects/phpseclib/files/phpseclib1.0.19.zip/download

The *.zip file is 274KB.

> `Call to undefined function ssh2_connect()`
> 
> Do I need to edit PHP.ini?

That's why I created phpseclib - because trying to do SFTP in PHP is otherwise a pain in the ass, as you're kinda figuring out.

Anyway, if you're on a host using cPanel then it sounds like it could very well be a shared host and, in that scenario, you can't install extensions that aren't already installed. I mean, normally, you'd install the ssh2 extension by doing `pecl install ssh2` and then editing the php.ini, but you're probably going to need sudo access to do `pecl install` and, given your use of cPanel, my guess would be that you don't have sudoer access.

You're only other choices would then be using OpenSSH on the CLI. eg. `exec('sftp ...');` but that opens up it's own can of worms. `exec` may be a disabled function and, even if it isn't, passing in `$_REQUEST` variables to the command line can introduce vulnerabilities if you're not careful.

Anyway, if you don't want to use phpseclib, then I recommend you post on stackoverflow. Honestly, I'd even recommend posting on stackoverflow _for_ phpseclib support as stackoverflow has waaaay more eyes than the phpseclib GitHub project does. But since you're not wanting to use phpseclib, then that makes my recommendation of posting your question stackoverflow that much more prudent.

**Comment by David263** [comment]
As a matter of fact, although I use cPanel on my remote server (but not on my local server, which is under Windows), my hosting is a VPS, not shared. Does that change your recommendation? I want to use sFTP on my local Windows computer, not on my cPanel remote server. I know these things are complicated!

I can install most standard Apache and PHP extensions, I think, by uncommenting them in the config files, but phpseclib is not found in either Apache or PHP. I have never run "pecl" and I just tried it and it is not a Windows command.

I'm willing to use phpseclib, but I need more help. Perhaps it needs a tutorial for Windows?

I am able to use the command line just fine, since this is my local computer and it has no incoming WAN access. But I will need help to use OpenSSH since my public key was allocated by cPanel. cPanel gave me two files: x.key starts "-----BEGIN RSA PRIVATE KEY-----" and x.ppk starts "PuTTY-User-Key-File-2: ssh-rsa".

Again, I have little experience with security, so I do need help, probably like hundreds of other developers trying to do the same thing.

I would use StackOverflow, but they have banned me from asking questions because they don't like my questions (as a software engineer with 40 years of experience, I think my questions are fine). They currently let me ask one question every six months.

**Comment by terrafrost** [comment]
I mean, keep in mind that I'm the author of phpseclib whereas I'm not the author of libssh2, so I'm probably a bit biased in phpseclib vs libssh2.

> I'm willing to use phpseclib, but I need more help. Perhaps it needs a tutorial for Windows?

I'd [install Composer for Windows](https://getcomposer.org/doc/00-intro.md#installation-windows) and then, in the directory of your choosing, do, via the CLI, `composer require phpseclib/phpseclb:~2.0`. Then, at the top of the PHP file you want to use phpseclib in, do `require __DIR__ . '/vendor/autoload.php';`.

To do libssh2 in Windows you'd need to download the DLL from http://pecl.php.net/package/ssh2

That said, I'd recommend getting used to Composer as that's the way a lot of libraries are going these days.

> cPanel gave me two files: x.key starts "-----BEGIN RSA PRIVATE KEY-----" and x.ppk starts "PuTTY-User-Key-File-2: ssh-rsa".

phpseclib supports both formats, libssh2 only supports the first one.

See https://phpseclib.com/docs/why#phpseclib-vs-libssh2 for more reasons to use phpseclib over libssh2!

(as I said earlier, I'm a bit biased towards phpseclib lol)

**Comment by David263** [comment]
Sorry to take so much of your time. I wish I could find a small library that contains only the functions needed for ordinary sFTP work, using just standard RSA encryption. The libssh2 extension, just like phpseclib, contains all sorts of other stuff.

Anyway, I can't seem to find any simple instructions for downloading your extension into the php/ext folder so it can be used with PHP. I really and truly do not want to use PEAR, PECL, Chocolatey, or Composer, none of which work in the standard Windows way, and if they are anything like npm, might use an enormous amount of disk space .

At this point I  just want to download phpseclib.dll, enable it in PHP.ini, and then follow your documentation for use. Please help me find the file?

**Comment by Sxderp** [comment]
I don't believe there is a .dll file associated with phpseclib. Pretty sure everything is built in pure PHP.

What you want is to download the 1.0.x release [here](https://github.com/phpseclib/phpseclib/releases/tag/1.0.19) (extract the zip). And then "include" / "require" the files with the classes / functions that you want.

Something like:
```php
<?php
set_include_path(get_include_path() . PATH_SEPARATOR . 'phpseclib');
include('Net/SFTP.php');

// Rest of your code.
?>
```

Check out the 1.0.x documentation on sourceforge:
http://phpseclib.sourceforge.net/
http://phpseclib.sourceforge.net/sftp/examples.html#put

----

With this said. I feel like the entire thread is almost a case of the [XY Problem](https://meta.stackexchange.com/questions/66377/what-is-the-xy-problem). If you just want to copy files, or copy files recursively, is a call to `scp` insufficient? The tool is available on all (up-to-date) operating systems (including Windows).

**Comment by David263** [comment]
All PHP extensions in windows are DLL files. Is phpseclib only for Linux? I don't want to use PuTTY or WinSCP because I want to conditionally upload programs using a PHP program.

I will try your suggested code, thanks. Sorry to bloat the Issues page with all these postings, but I don't know how to contact you more directly.

**Comment by Sxderp** [comment]
> All PHP extensions in windows are DLL files.

phpseclib is not an "extension" in the traditional sense. It it a collection of pre-defined functions and classes that you can include directly in your project directory. As a result you do not have to modify the php configuration (no .dll to install / no change to php.ini). You just download / copy the files into your project.

**Comment by David263** [comment]
Thank you for the download link. So, now I know that this library is in PHP source code, and takes up 1.24 MB on disk. That is big. Is there a subset that just does sftp? I can't imagine needing 1.24 MB of code to upload a file securely. The file SFTP.php contains only 3244 lines, and is only 101 KB in size. Can I just use SFTP.php alone?

**Comment by David263** [comment]
The example at the beginning of SFTP.php logs in with a password. Is there an example of logging in with an encrypted private key from a file?

**Comment by terrafrost** [comment]
> So, now I know that this library is in PHP source code, and takes up 1.24 MB on disk. That is big. Is there a subset that just does sftp? I can't imagine needing 1.24 MB of code to upload a file securely. The file SFTP.php contains only 3244 lines, and is only 101 KB in size. Can I just use SFTP.php alone?

Assuming you're using phpseclib 1.0 or 2.0, you can delete:

- File/*.php
- Net/SSH1.php
- Net/SCP.php
- Net/SFTP/Stream.php
- System/*.php

That'll knock it down to 922KB from 1.27MB. That may still be larger than you'd like it to be but trying to make it any smaller than that will depend on your SFTP server. SFTP _uses_ SSH so you do need SSH2.php.

Here's how SSH works. The various crypto algorithms are first negotiated (key exchange, server signature verification, symmetric key encryption, hash algorithm, etc). Once this is done key exchange is performed using the negotiated algorithm so that both sides can have the same symmetric key. This requires Math/BigInteger.php since that'll make use of > 64-bit numbers. Then, if you're logging in with a private key, Crypt/RSA.php is required. RSA makes use of > 64-bit numbers so you'll once again need Math/BigInteger.php. The files required for symmetric key crypto ultimately depend in the symmetric key crypto algorithm that was negotiated. It could be If it's AES then you'll need Crypt/AES.php, Crypt/Base.php and Crypt/Rijndael.php which means that the following could be deleted:

- Crypt/Blowfish.php
- Crypt/DES.php
- Crypt/RC2.php
- Crypt/RC4.php
- Crypt/TripleDES.php
- Crypt/Twofish.php

That'd free up 193KB. But that's only assuming that AES is being used and I'd need to see the output of `$ssh->getAlgorithmsNegotiated()` to know that. And even if that worked for you it might not work for other people, so it will never be removed from the main phpseclib. Like even if phpseclib were eventually split out into separate packages for each "component" _all_ the symmetric key algorithms would be dependencies.

The fact that the RSA key is encrypted throws another wrench into things. How is that encrypted? Keys starting off with `PuTTY-User-Key-File-2: ssh-rsa` _only_ support AES encryption but keys starting off with `-----BEGIN RSA PRIVATE KEY-----`...  phpseclib 1.0.x supports DES, TripleDES and AES for encryption on those.

Now, if all you're doing is wanting to upload a file, then there are definitely functions you could delete. You don't need the `exec` method in SSH2.php. I guess you wouldn't need the `get` method in SFTP.php either.

tldr yes it's def possible to make phpseclib smaller but the smallest I'd advise making it is 922KB.

I suppose you could also use https://php-minify.com/ as well but that tool could have it's own issues and I'm not going to provide support for a version of phpseclib minified with that (as the odds are that the problem would be with that tool and not phpseclib itself)

**Comment by David263** [comment]
BigInteger.php is only  20.9 KB, so it is not a problem. Anyway, elsewhere here you say that without some additional libraries phpseclib is very slow.

At this point, I'm looking into running the native Windows command sftp using php exec function (but I can't find instructions for it), and installing the libssh2.dll extension from https://pecl.php.net/package/ssh2/1.2/windows (but I don't know which of its many versions to install).

Thank you for taking the time to work with me in such detail. I really appreciate it.

**Comment by Sxderp** [comment]
> Is there an example of logging in with an encrypted private key from a file?

SFTP extends the SSH interface. So you should be able to use the same code as in the SSH example. I have not tested this as I don't have a php environment readily available.

```php
<?php
include 'Crypt/RSA.php';
include 'Net/SSH2.php';

$key = new Crypt_RSA();
$key->setPassword('whatever');
$key->loadKey(file_get_contents('privatekey'));

$ftp = new Net_SFTP('www.domain.tld');
if (!$ftp->login('username', $key)) {
    exit('Login Failed');
}

$ftp->put('remote_file.ext', file_get_contents('local_file.ext'));
```

---

Now, if you really don't want to use phpseclib, I don't there's much more that can be said. It might be better migrating the request to stackoverflow (might even find similar questions).

**Comment by Sxderp** [comment]
> At this point, I'm looking into running the native Windows command sftp ...

I'm not sure about a native Windows SFTP command, but there is a SCP command. If your remote server supports SFTP then it likely supports SCP too.

The command arguments are (basically) the same as they are on Linux, here is the [man page](https://linux.die.net/man/1/scp). Something like the following could work in PHP. However, I don't know how you would put in / store the private key passphrase. You would need one without a passphrase.

```php
<?php
$keyfile = 'privatekey';
$localfile = 'C:\Users\Default\Desktop\resource.txt';
$remotedirectory = '/home/default/projects/';

exec("scp -i \"${keyfile}\" \"${localfile}\" \"${remotedirectory}\"");
```

**Comment by David263** [comment]
It turns out that there is an scp on my Windows. Here is the Help:
usage: scp [-346BCpqrv] [-c cipher] [-F ssh_config] [-i identity_file]
           [-l limit] [-o ssh_option] [-P port] [-S program] source ... target
What is "cipher", "ssh_config", "identity_file", etc.? Where is the documentation?

**Comment by David263** [comment]
Your example has 'include "crypt/RSA.php";'.
When I try this, I get this error message:
"Warning: include(crypt/RSA.php): failed to open stream: No such file or directory"
I really need examples that work.

**Comment by Sxderp** [comment]
#### Using phpseclib

>  "Warning: include(crypt/RSA.php): failed to open stream: No such file or directory"

1. Make sure you have the 'phpseclib' folder in your working directory. This folder has other folders like "Crypt", "Net", etc
2. Make sure you set the include path (my previous example).
```php
set_include_path(get_include_path() . PATH_SEPARATOR . 'phpseclib');
```

----

#### Using native scp / ssh
> What is "cipher", "ssh_config", "identity_file", etc.? Where is the documentation?

You only need "identify_file". This is the path to your RSA private key. Again, if your key is encrypted (has a passphrase) you'll have to input it. I do not know how to do this in a PHP program. You'll likely need a key without a passphrase. There are ways to convert, but that is out of scope of this thread.

On your command line:
```
scp -i C:\Users\Default\Desktop\RSA_Private_Key C:\Users\Default\Desktop\Local_File username@server.tld:/home/default/projects
```

Or for recursive transfers (must specify a local directory):
```
scp -r -i C:\Users\Default\Desktop\RSA_Private_Key C:\Users\Default\Desktop\Local_Directory username@server.tld:/home/default/projects
```

Note: If you have spaces in your file names then the paths must be quoted.

> Where is the documentation?

The man page I linked earlier for `scp` should be sufficient. Even though you're on Windows the arguments are the same. For more information you can follow the links to the `ssh` man page. Again, the command and arguments are the same on Windows and Linux.

**Comment by terrafrost** [comment]
One thing to keep in mind with SCP is that it does require shell access, which not all servers provide. See https://github.com/phpseclib/phpseclib/blob/2.0.30/phpseclib/Net/SCP.php#L141 to see specifically how SCP works.

Overall, I think SFTP is the far superior protocol, which is why I dropped SCP support from 3.0.

**Comment by David263** [comment]
I can use "exec" and I have full access to my own remote server. I don't mind using scp if it works, but I don't as yet have a working example of code to upload files. I have no interest in downloading files, and I would be fine with sftp if I had working code that didn't require tons of hidden code to be installed. So far the examples I have of code don't work.

Here is an example:

`exec("scp -i $ppkPath test.txt $user@$host 2> x.txt");`

This code produces no error message, fails to upload the file, and produces a file "user@host" in the local directory that contains the contents of test.txt .

I am able to upload files manually using a standard sftp program (coreFTP) and a standard shell program (PuTTY), but I need to do uploads from PHP as part of a website building program. I have been trying to get help at Experts Exchange, but the experts there have not been helping very much so far.

**Comment by Sxderp** [comment]
Your missing a very important colon `:` between `$user` and `$host`.

Also, I do not believe you will be able to use the `ppk` file. You will need the `key` (one that starts with RSA PRIVATE KEY).

**Comment by David263** [comment]
I changed my code to 

`exec("scp -i $keyPath test.txt $user:$host 2> x.txt");
`but got the following error message:

ssh: Could not resolve hostname $user: Name or service not known
lost connection

**Comment by Sxderp** [comment]
Sorry, if you're reading my previous message there was an error. I had edited it as soon as I could. Please check the updated text on Github.

**Comment by Sxderp** [comment]
Here is a full phpsec lib example (tested it).
```php
set_include_path(get_include_path() . PATH_SEPARATOR . 'phpseclib');
include 'Crypt/RSA.php';
include 'Net/SFTP.php';

$key = new Crypt_RSA();
$key->setPassword('testtest');
$key->loadKey(file_get_contents('key'));

$ftp = new Net_SFTP('math-gateway0');
if (!$ftp->login('sdaniele3', $key)) {
    exit('Login Failed');
}

$ftp->put('/home/sdaniele3/local.txt', file_get_contents('local.txt'));
```
```bash
# working directory
$ ls 
example.php  key  key.pub  local.txt  phpseclib

# phpseclib folder
$ ls phpseclib/
bootstrap.php  Crypt  File  Math  Net  openssl.cnf  System

# run command
$ php example.php


# Then on the server
$ ls -l /home/sdaniele3/local.txt 
-rw-rw----. 1 sdaniele3 nobody 5 Dec 29 13:21 /home/sdaniele3/local.txt
```

Here is a full scp example (tested it). Your going to have to enter the passphrase for every transfer or unencrypted the key.
```php
<?php
$user = 'sdaniele3';
$server = 'math-gateway0';
$path = '/home/sdaniele3/';
$keyfile = 'key';
$file = 'local.txt';

exec("scp -i $keyfile $file $user@$server:$path");
```
```bash
# working directory
$ ls 
example.php  key  key.pub  local.txt

# run command
$ php example.php 
Enter passphrase for key 'key': 


# Then on the server
$ ls -l /home/sdaniele3/local.txt 
-rw-rw----. 1 sdaniele3 nobody 5 Dec 29 13:30 /home/sdaniele3/local.txt
```

**Comment by David263** [comment]
I tried again with

`exec("scp -i $keyPath test.txt $user@$host:/public_html 1> x.txt 2>&1");`

and got this output:

```
ssh: connect to host 162.254.253.20 port 22: Connection refused
lost connection
```

Please note that the port number (which is not 22 for my remote server) is not included in the command.

**Comment by Sxderp** [comment]
Port 22 is the default ssh port. To change the port, as listed in the [man page](https://linux.die.net/man/1/scp) you can add `-P PORT` to the command.

ex:
`scp -P PORT ...`

**Comment by David263** [comment]
Thank you! This time the error message is:

```
Host key verification failed.
lost connection

```

**Comment by David263** [comment]
And the key file begins:

```
-----BEGIN RSA PRIVATE KEY-----
Proc-Type: 4,ENCRYPTED
DEK-Info: AES-128-CBC,7351603C5596CD4AE349F16A1478169A

```
I looked at the man page but cannot see how to specify the key encryption password (usually called a passphrase).

**Comment by Sxderp** [comment]
Ah. Host keys are used to identify servers. When you connect to a server the host presents the key. On the very first connection your supposed to manually verify it (getting an offline copy) and then accepting it. In reality almost nobody does this and just accepts the key. /shrug.

You have two options.
1. Run the scp command outside of PHP so that the scp program prompts you to accept the key, and then accept it.
2. Add an argument to ignore the key. `-o StrictHostKeyChecking=no`

> I looked at the man page but cannot see how to specify the key encryption password (usually called a passphrase).

I don't think there's a way. You will need to decrypt the key or input the password when prompted.

**Comment by David263** [comment]
Does "host key" really mean the password I originally used to create x.ppk? Or is it some password that the host computer supplies? I'm confused.

**Comment by David263** [comment]
I tried

`echo $ppkPwd | exec("scp -P $securePort -i $keyPath test.txt $user@$host:/public_html 1> x.txt 2>&1");
`
And the error message was

```
Host key verification failed.
lost connection

```

**Comment by Sxderp** [comment]
Unfortunately there's some bad terminology. Some people refer to the public / private keys (`.ppk` and `.key`) as "host keys". While they're not wrong, it is somewhat misleading.

In general I believe they refer to the following:

1. Host key. A cryptographicly secure public / private keypair used to identify a server.
2. Identify file. A cryptographicaly secure public / private keypair used to identify a user.
3. Passphrase. A phrase used to decrypt an encrypted identify file.

Do note that there are multiple different types of 1 and 2. RSA/DSA/ECDSA/etc. And theoretically can be used interchangeably.

---

I suggest running the `scp` command directly on the Windows command line. Once you have a working command then migrate it into PHP. That way you can see all the error messages and prompts.

**Comment by Sxderp** [comment]
Alternatively, if you just use phpseclib you won't have to deal with unverified host keys nor being unable to input the private key passphrase. You can do that all in the code.

A 1MB library size is a small price to pay for a complete PHP only solution (little / no outside dependencies). Full, working examples have been posted in this thread.

**Comment by David263** [comment]
Okay, I supplied all the secret information and ran the bare scp command on the Windows command line.

Here are the messages. What do they mean?

```
The authenticity of host '[162.254.253.20]:*** ([162.254.253.20]:***)' can't be established.
ECDSA key fingerprint is SHA256:KOSlZ10m6Nid/2ar0Mgaga0ckv1CEImmiWGe6ZdDbEU.
Are you sure you want to continue connecting (yes/no)?
Please type 'yes' or 'no':
Warning: Permanently added '[162.254.253.20]:***' (ECDSA) to the list of known hosts.
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
@         WARNING: UNPROTECTED PRIVATE KEY FILE!          @
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
Permissions for 'C:/WAMP/certificates/A2-nsr.key' are too open.
It is required that your private key files are NOT accessible by others.
This private key will be ignored.
Load key "C:/WAMP/certificates/A2-nsr.key": bad permissions
nsr@162.254.253.20: Permission denied (publickey,gssapi-keyex,gssapi-with-mic).
lost connection

```

I ran the PHP program again and got:

```
Host key verification failed.
lost connection

```

**Comment by Sxderp** [comment]
I'm not sure the exact steps to fix that permission issue, since you're using a Windows machine. The basics is "make sure you, and only you" has access to the file. Key files are "identifies" therefore the only person that should access to the "identity" is the owner. The error message you are receiving is a sanity check saying "HEY OTHER PEOPLE COULD IMPERSONATE YOU!" (should they have access to the machine).

**Comment by David263** [comment]
```
The authenticity of host '[162.254.253.20]:*** ([162.254.253.20]:***)' can't be established.
ECDSA key fingerprint is SHA256:***.
Are you sure you want to continue connecting (yes/no)?
Please type 'yes' or 'no':
Warning: Permanently added '[162.254.253.20]:***' (ECDSA) to the list of known hosts.
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
@         WARNING: UNPROTECTED PRIVATE KEY FILE!          @
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
Permissions for '***.key' are too open.
It is required that your private key files are NOT accessible by others.
This private key will be ignored.
Load key "***.key": bad permissions
***@162.254.253.20: Permission denied (publickey,gssapi-keyex,gssapi-with-mic).
lost connection

```
I ran the PHP program again and got:

```
Host key verification failed.
lost connection
```

**Comment by David263** [comment]
I feel that I am too close to give up now. If I can get this working I can help hundreds of other people who want to do the same thing. I agree that phpseclib is an alternative.

**Comment by David263** [comment]
I found a Windows solution to this particular error at https://superuser.com/questions/1296024/windows-ssh-permissions-for-private-key-are-too-open and I am working on it.

**Comment by terrafrost** [comment]
I'm going to close this topic. I believe y'all should still be able to comment in it and get / provide support to one another but, at this point, since phpseclib isn't being utilized, it's not exactly an outstanding bug report / ticket.

If y'all are unable to post I guess shoot me an email and I can re-open it or something.

Thanks and good luck!
